src/caf/toolkit/cost_utils.py

```diff
@@ -141,6 +149,46 @@ def band_share_vals(self) -> np.ndarray:
         trip_vals = self.trip_vals
         return trip_vals / np.sum(trip_vals)
 
+    @staticmethod
+    def calculate_weighted_averages(
+        matrix: np.ndarray, cost_matrix: np.ndarray, bin_edges: list[float] | np.ndarray
+    ):
+        """
+        Calculate weighted averages of bins in a cost distribution.
+
+        Parameters
+        ----------
+        matrix: np.ndarray
+            The matrix to calculate the cost distribution for. This matrix
+            should be the same shape as cost_matrix
+
+        cost_matrix: np.ndarray
+            A matrix of cost relating to matrix. This matrix
+            should be the same shape as matrix
+
+        bin_edges: list[float] | np.ndarray
+            Defines a monotonically increasing array of bin edges, including the
+            rightmost edge, allowing for non-uniform bin widths. This argument
+            is passed straight into `numpy.histogram`
+
+        Returns
+        -------
+        np.ndarray
+            An array to be passed into a dataframe as a column.
+        """
+        df = pd.DataFrame(
+            {
+                "cost": pd.DataFrame(cost_matrix).stack(),
+                "demand": pd.DataFrame(matrix).stack(),
+            }
+        )
+        df["bin"] = pd.cut(df["cost"], bins=bin_edges)
+        df["weighted"] = df["cost"] * df["demand"]
+        grouped = df.groupby("bin", observed=False)[["weighted", "demand"]].sum().reset_index()
+        grouped["bin_centres"] = grouped["bin"].apply(lambda x: x.mid)
+        grouped["averages"] = grouped["weighted"] / grouped["demand"]
+        return grouped["averages"].fillna(grouped["bin_centres"].astype("float")).to_numpy()
```

Comment:
Need to come back and think about this in more detail. Might be one worth chatting about the best way to do this